light interactive pr review

vilmibm opened the pull request:
This PR adds "light" interactive reviewing; ie still a top-level review but with prompts for review type and adding a body via your editor.

![review](https://user-images.githubusercontent.com/98482/81439901-c2c15b00-9134-11ea-9361-e0d54868498c.gif)



Fixes #434

billygriffin commented on api/queries_pr.go:
```diff
@@ -6,10 +6,24 @@ import (
+
+type PullRequestReviewState int
+
+const (
```
Testing commenting on reviews

vilmibm commented on command/pr_review_test.go:
```diff
@@ -112,26 +112,17 @@ func TestPRReview_number_arg(t *testing.T) {
-	initBlankContext("", "OWNER/REPO", "master")
+	initBlankContext("", "OWNER/REPO", "feature")
 	http := initFakeHTTP()
 	http.StubRepoResponse("OWNER", "REPO")
```
test

mislav commented on command/pr_review.go:
```diff
@@ -128,10 +134,128 @@ func prReview(cmd *cobra.Command, args []string) error {
+		{
+			Name: "confirm",
+			Prompt: &survey.Select{
+				Message: "What's next?",
+				Options: []string{
+					"Submit",
+					"Cancel",
```
As I understand, right now this flow has 3 steps:
1. “What kind of review do you want to create?”
  - Comment
  - Approve
  - Request changes
  - Cancel
2. “Review body” (opens text editor)
3. “What's next?”
  - Submit
  - Cancel

Have we considered trying to get away with just 2 prompts?

1. “Review body” (opens text editor)
2. “What's next?”
  - Comment
  - Approve
  - Request changes
  - Cancel

I think this would be both simpler for our users and require less code. /cc @ampinsk

ampinsk replied:
Thanks for the suggestion @mislav! I find conflating the submit step with choosing the type of review to be a little confusing. I think giving them their own prompts while slightly more steps, makes more semantic sense. 

I was also thinking for this we could do just a simpler `Submit? [Y/n]` instead of `What's next?`

vilmibm replied:
I switched to the Y/n prompt

ampinsk commented on command/pr_review.go:
```diff
@@ -125,10 +133,123 @@ func prReview(cmd *cobra.Command, args []string) error {
+					"Comment",
+					"Approve",
+					"Request Changes",
+					"Cancel",
```
Could we nix the `Cancel` option in the first prompt? I think this muddies up that question and you can cancel by saying no at the submit step.

ampinsk commented on command/pr_review.go:
```diff
@@ -125,10 +133,123 @@ func prReview(cmd *cobra.Command, args []string) error {
+		{
+			Name: "reviewType",
+			Prompt: &survey.Select{
+				Message: "What kind of review do you want to create?",
```
```suggestion
				Message: "What kind of review do you want to give?",
```
This flowed a little better to me? Thoughts?

ampinsk commented on command/pr_review.go:
```diff
@@ -125,10 +133,123 @@ func prReview(cmd *cobra.Command, args []string) error {
+				Options: []string{
+					"Comment",
+					"Approve",
+					"Request Changes",
```
```suggestion
					"Request changes",
```
Been trying to stick to sentence case like GitHub.com :)

ampinsk commented on command/pr_review.go:
```diff
@@ -125,10 +133,123 @@ func prReview(cmd *cobra.Command, args []string) error {
+			Prompt: &surveyext.GhEditor{
+				EditorCommand: editorCommand,
+				Editor: &survey.Editor{
+					Message:  "Review Body",
```
```suggestion
					Message:  "Review body",
```

tierninho commented:
- Do we want affirmation when a user selects `Cancel` or selects `No` when the submit prompt is shown? I noticed it isn't present here but we show "Discarding."  when a user abandons `gh issue create`.
- If a user does this command `gh pr review 13 -b test` on their own PR, we should immediately show the `failed to create review: Can not approve your own pull request` and not wait until the user goes through the steps. Note, we show the error immediately when the `-a` flag is used.
- For both the `-c` and `-r` flags, I got "accepts at most 1 arg(s), received 2". Is there any example I can follow?
- The user can skip the comment section, but ultimately needs to fill it out else they have to start over. Is this because we cannot easily strip out `enter to skip` option?
- Are we supporting line breaks? 
![Screen Shot 2020-05-08 at 2 18 07 PM](https://user-images.githubusercontent.com/14828183/81458590-d53c9400-9136-11ea-80b8-6dc39ed19a60.png)
- I attempted to use `-b` but it didn't seen to carry through the the body of the comment or review. Any tips?
- An affirmation message would be great if a PR was successfully approved or commented or reviewed.
- How is a review different than a comment?

mislav commented on command/pr_review.go:
```diff
@@ -125,10 +133,123 @@ func prReview(cmd *cobra.Command, args []string) error {
+			return nil, err
+		}
+
+		fmt.Fprintf(out, "Got:\n%s", renderedBody)
```
Was printing the preview of the review body intended in the final product? We don't do that in `issue/pr create`, so doing it here looks odd. Also, the user has just seen this text because they've authored it in their text editor, so what would be the intended purpose of printing it out right after?

vilmibm replied:
we've discussed doing it for pr/issue as a review step. I thought this was a good opportunity to try it out.

vilmibm commented:
@probablycorey I updated `GhEditor` to support a `BlankAllowed` property; if set to false, the "enter to skip" behavior is disabled.

vilmibm commented:
@tierninho 

- Do we want affirmation when a user selects Cancel or selects No when the submit prompt is shown? I noticed it isn't present here but we show "Discarding." when a user abandons gh issue create.

Good idea, added a discarding message.

- If a user does this command gh pr review 13 -b test on their own PR, we should immediately show the failed to create review: Can not approve your own pull request and not wait until the user goes through the steps. Note, we show the error immediately when the -a flag is used.

Fixed.

- For both the -c and -r flags, I got "accepts at most 1 arg(s), received 2". Is there any example I can follow?

Can you paste what you ran? sounds like you were missing the `-b`?

- The user can skip the comment section, but ultimately needs to fill it out else they have to start over. Is this because we cannot easily strip out enter to skip option?

Fixed

- Are we supporting line breaks?

Yes, but it's markdown-rendered. Two adjacent lines in markdown are joined into one paragraph. A blank line is needed between them.

- I attempted to use -b but it didn't seen to carry through the the body of the comment or review. Any tips?

`-b` is only to be used in conjunction with `-a/-c/-r`.

- An affirmation message would be great if a PR was successfully approved or commented or reviewed.

Fixed

- How is a review different than a comment?

They are two separate concepts on github.com.

vilmibm commented:
I've taken care of all the feedback and none of the new code is very involved so I'm going to go ahead and merge~

tierninho commented:
Thanks for all the fixes!

> Can you paste what you ran? sounds like you were missing the -b?

Yes, I was not including the `-b`, so we can ignore.
